### ISSUE TITLE
spaces-renamer doesn't rename desktops on Big Sur

### ISSUE THREAD

**Issue by bomtailey** [dado3212/spaces-renamer#67]
Running Big Sur 11 Beta (20A5364e). Followed the installation steps and everything seems to have worked properly. The spaces-renamer icon shows on the menu bar and when clicked displays the prompt to rename the desktops however they don't get renamed. SIP is disabled.

**Comment by NikolaiKunewa** [comment]
Have the same problem...Really though I had this one. Appreciate if you make a quick tutorial

**Comment by itsrainingdata** [comment]
Same issue on Catalina 10.5.6.

**Comment by dado3212** [comment]
Can you try updating to v1.9.0, and also deleting the .bundle from MacForge, and installing by dragging and dropping?

![image](https://user-images.githubusercontent.com/8919256/93972317-e2919e00-fd26-11ea-8b2b-0e9fbf849ed1.png)

Selecting spaces-renamer, and clicking the trash can, and then dragging and dropping should work.

**Comment by dado3212** [comment]
RE: "everything seems to have worked properly" the bit where the icon shows up and allows you to type in names doesn't need SIP disabled, or MacForge running, so it's probably an issue with one of those two. Can you paste a screenshot of what shows up when you run `csrutil status`?  And paste a screenshot of your full menu bar options?

**Comment by itsrainingdata** [comment]
I believe the issue was that the .bundle did not install correctly: Neither dragging and dropping nor installing directly from MacForge worked for me. I managed to get it working by manually copying and pasting the .bundle to /Library/Application Support/MacEnhance/Plugins.

Am now running v1.9.0 without issues.

**Comment by jackkrone** [comment]
Had the same issue described. Manually dragging and dropping worked for me as well. Thanks @itsrainingdata and @dado3212

**Comment by itsrainingdata** [comment]
Update: I am encountering some quirks with renaming spaces (e.g. alignment of names w/ multiple monitors, failure to rename some spaces, issues with rearranging spaces). 

If the issues continue to persist, I will open up a separate issue (unless @dado3212 wants to consolidate everything here).

**Comment by bomtailey** [comment]
Tried drag and drop with version 1.9.0 but get the same result (i.e renaming doesn’t happen)

**Comment by dado3212** [comment]
@itsrainingdata open a new task if it's a new issue!

@w0lfschild any thoughts on MacForge interface?

**Comment by w0lfschild** [comment]
@dado3212 I've gone ahead and updated the version on MacForge to 1.9.0

**Comment by bomtailey** [comment]
@dado3212 I would still (very much) like to get this capability working. I've tried dragging and dropping into /Library/Application Support/MacEnhance/Plugins as suggested but no luck. I've read that it doesn't work with Big Sur, maybe that's the problem. The response I get from issuing csrutil status is attached. The screen shot of my menu bar is attached.  BTW, where it was suggested to install it by dragging and dropping the bundle, I don't know where the destination is. I used the suggestion of @itsrainingdata.
![csrutil status result](https://user-images.githubusercontent.com/17606064/95261465-27f1a900-07c6-11eb-808e-6cb4928052c4.png)
![menu bar screenshot](https://user-images.githubusercontent.com/17606064/95261471-29bb6c80-07c6-11eb-8f94-152cf7fbabb7.png)

**Comment by bomtailey** [comment]
Anybody have this working with Big Sur Beta 11.0 (20A5384c)?

**Comment by w0lfschild** [comment]
> Anybody have this working with Big Sur Beta 11.0 (20A5384c)?

This won't work on any of the latest Big Sur releases without changes to MacForge to prevent the Dock from crashing.

Any plugin that loads into the Dock will crash loop it on Beta 7+

**Comment by bomtailey** [comment]
@w0lfschild thanks for that clarification,

**Comment by itsrainingdata** [comment]
@w0lfschild Does this mean spaces-renamer/MacForge fundamentally can't work on Big Sur? Or just that either/or need to be updated to preserve functionality?

**Comment by w0lfschild** [comment]
@itsrainingdata 

> @w0lfschild Does this mean spaces-renamer/MacForge fundamentally can't work on Big Sur? Or just that either/or need to be updated to preserve functionality?

No it doesn't mean mean spaces-renamer or MacForge fundamentally can't work on Big Sur. 

It means exactly what I said.

> Any plugin that loads into the Dock will crash loop it on Beta 7+
> This won't work on any of the latest Big Sur releases **without changes to MacForge to prevent the Dock from crashing**.

I have spaces-renamer running on beta 10 and it works just fine (with a fix for MacForge).

Specifically the crash is related to loading `NSApplication`, which happens if you have basically any framework (AppKit, Cocoa, Foundation, etc..) linked in the plugin.

```
67020474 bear trap: NSApplication initialized in Dock. Please log a bug.

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   com.apple.AppKit              	0x00007fff22c9b1ad +[NSApplication load] + 125
```

@JSLegendre has already written a fix but I have not released anything yet.

**Comment by jslegendre** [comment]
Took a quick look through the code here and the only part of the plugin that uses anything from AppKit is the method `- (BOOL)probablyDesktopSwitcher:(CGRect)rect`.  

To get this working with MacForge on Big Sur (no 'fix' necessary), replace `@import AppKit;` (line 9) with `@import Foundation;`. 

Then replace (line 395)
```  NSArray *const screenArray = [NSScreen screens];
  for (int i = 0; i < screenArray.count; i++) {
    NSScreen *const screen = [screenArray objectAtIndex:i];
    if (screen.visibleFrame.size.width == rect.size.width) {
      return true;
    }
  }
```
With this
```
  CGDirectDisplayID displayArray[5];
  uint32_t displayCount;
  CGGetActiveDisplayList(5, displayArray, &displayCount);
  
  for (int i = 0; i < displayCount; i++) {
    if(CGDisplayPixelsWide(displayArray[i]) == rect.size.width)
      return true;
  }
```

**Comment by bomtailey** [comment]
Did anyone ever get this working with Big Sur? I made the changes suggested by Jeremy LeGendre and was able to produce a bundle but no executable for what reason I don’t know. Any plans to upgrade?
Thanks

**Comment by jslegendre** [comment]
@bomtailey Did you select the right scheme in Xcode?

**Comment by BuenoGrande** [comment]
> Can you try updating to v1.9.0, and also deleting the .bundle from MacForge, and installing by dragging and dropping?
> 
> ![image](https://user-images.githubusercontent.com/8919256/93972317-e2919e00-fd26-11ea-8b2b-0e9fbf849ed1.png)
> 
> Selecting spaces-renamer, and clicking the trash can, and then dragging and dropping should work.

Hello I tried using this method but it doesn't seem to work. Now I cannot even launch the SpacesRenamer app, even though I dragged and dropped it in MacForge. After installing it in MacForge, when trying to run the app Apple doesn't allow me to launch the app. When I go to the settings in security and press "open it anyway", I have an error: OSStatus error 99999.


I also disabled csrutil and checked its status.. Do you have any fixes? I really love your app.

**Comment by dado3212** [comment]
@BuenoGrande that sounds like a different problem, can you open a new issue?

**Comment by dado3212** [comment]
Huge thanks to @jslegendre for the suggested fix. I think it'll probably be resolved in a MacForge fix soon, but I incorporated it into v1.9.1 if anyone's chomping at the bit. Confirmed it works with multiple monitors on Catalina, don't have a Big Sur device to confirm, but I trust @jslegendre.

For @bomtailey, @NikolaiKunewa, @itsrainingdata, see if v1.9.1 resolves this issue, closing the task for now. If it's still not working, feel free to reopen this, or file a new issue.
